### PR TITLE
Add hash checks for `drep registration-certificate` and `drep update-certificate`

### LINKED ISSUE #882: [FR] - Add File Hash Validation when Building Transaction
**Internal/External**
*External*

**Area**
*Other* 

**Describe the feature you'd like**
When publishing an action that expects or includes a remotely hosted metadata file and a confirmation hash (e.g. pool registration/update, governance action creation/submission) and using `cardano-cli transaction build` then we should provide a mechanism to provide some dummy-proofing for the user w.r.t. validating that the provided hash matches the file contents at the remote source.

**Describe alternatives you've considered**
The current alternative is for users to download the file themselves from the remote source and then hash the file and double-check that the values entered into these commands (`cardano-cli transaction build` or `cardano-cli governance create-info`) match. 

**Additional context / screenshots**
A very "costly" real world example is the first Governance Action published to mainnet: https://cexplorer.io/tx/15f82a365bdee483a4b03873a40d3829cc88c048ff3703e11bd01dd9e035c916/governance#data

File URL: [ipfs://QmWjcHsrq9kKHZZ7aPPFjqN6wLuxH9d8bcqssmrE7H4cvb](https://ipfs.io/ipfs/QmWjcHsrq9kKHZZ7aPPFjqN6wLuxH9d8bcqssmrE7H4cvb)

Here we had two potential points where the CLI could have provided confirmation/validation of the URI and the hash.

There are currently two types of certificates that rely on remotely hosted files and their hashes being published to the blockchain:
* Governance Actions
* Governance Votes
* dRep Registrations/Updates
* Stake Pool Registrations/Updates

**Governance Action Create**

```shell
dev@null:~$ body_hash=$(./cardano-signer hash --cip100 --data-file govaction.jsonld)
dev@null:~$ file_hash=$(b2sum -l 256 govaction.jsonld)
dev@null:~$ cardano-cli conway governance action create-info \
>  --mainnet \
>  --governance-action-deposit 100000000000 \
>  --deposit-return-stake-verification-key-file govaction.staking.vkey \
>  --anchor-url ipfs://QmWjcHsrq9kKHZZ7aPPFjqN6wLuxH9d8bcqssmrE7H4cvb \
>  --anchor-data-hash $body_hash \
>  --out-file governance.action
```

Here we have an opportunity for the CLI (assuming there is an `IPFS_GATEWAY_URI` environment variable set) to fetch the resource to a local temporary file and confirm that the correct and matching hash has been used. In the example shown, we've incorrectly used the `body_hash` instead of the `file_hash` (which is what actually happened with this first gov action).

**Governance Vote**
```shell
dev@null:~$ cardano-cli conway governance vote create \
> --yes \
> --governance-action-tx-id abc123 \
> --governance-action-index 0 \
> --drep-verification-key-file adam.drep.vkey \
> --anchor-url abc.123.fun \
> --anchor-data-hash abc123 \
> --out-file adam.vote
```

**dRep Registration Certificate**
```shell
dev@null:~$ cardano-cli conway governance drep registration-certificate \
> --drep-verification-key-file adam.drep.vkey \
> --key-reg-deposit-amt 2000000 \
> --drep-metadata-url abc.123.xyz \
> --drep-metadata-hash abc123 \
> --out-file adam.drep.cert
```

**dRep Update Certificate**
```shell
dev@null:~$ cardano-cli conway governance drep update-certificate \
> --drep-verification-key-file adam.drep.vkey \
> --key-reg-deposit-amt 2000000 \
> --drep-metadata-url abc.123.xyz \
> --drep-metadata-hash abc123 \
> --out-file adam.drep.cert
```

**Stake Pool Registration Certificate**

```shell
dev@null:~$ body_hash=$(./cardano-signer hash --cip100 --data-file govaction.jsonld)
dev@null:~$ file_hash=$(b2sum -l 256 govaction.jsonld)
dev@null:~$ pool_metadata=$(cardano-cli conway stake-pool metadata-hash --pool-metadata-file pool.json)
dev@null:~$ cardano-cli conway stake-pool registration-certificate \
> --mainnet \
> --stake-pool-verification-key-file mypool.cold.vkey \
> --vrf-verification-key-file mypool.vrf.vkey \
> --pool-pledge 100000000000 \
> --pool-cost 170000000 \
> --pool-margin 0.01 \
> --pool-reward-account-verification-key-file mypool.rewards.vkey \
> --pool-owner-stake-verification-key-file mypool.owner.vkey \
> --pool-relay-ipv4 192.168.0.1 \
> --pool-relay-port 1337 \
> --metadata-url https://github.com/crypto2099/mypool.json \
> --metadata-hash $file_hash \
> --out-file my-pool.cert
```

Here we have an opportunity for the CLI (maybe assuming there is an `IPFS_GATEWAY_URI` environment variable set if we switch to allowing IPFS URIs for stake pool metadata in the future) to fetch the resource to a local temporary file and confirm that the correct and matching hash has been used. In the example shown, we've incorrectly used the `file_hash` variable which is the `blake2b-256` of our Governance Action metadata file rather than our stake pool metadata file so this will cause issues and errors downstream with explorers (this has happened literally hundreds if not thousands of times in the 4 years since Shelley).

**Certificate Transaction Building**

When using `cardano-cli transaction build...` we can assume that the user is utilizing a "hot" network environment with a local or remote connection to a node and so, it should be possible to check when there is a `proposal-file`, `vote-file`, or `certificate-file` and their related metadata URLs and hashes are valid and either:
* Issue a warning if the hashes cannot be checked for some reason (file not accessible?)
* Issue a warning if the hash does not match the file contents

```
cardano-cli conway transaction build \
--mainnet \
--tx-in ${tx_in_id} \
--change-address ${my_change_address} \
--proposal-file ${gov_action_file} \
--vote-file ${gov_vote_file} \
--certificate-file ${pool_cert_file} \
--out-file my.tx.unsigned
```

### PR DESCRIPTION
**Pull request by palas**
# Changelog

```yaml
- description: |
    Added hash checks for `drep registration-certificate` and `drep update-certificate`
  type:
  - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  - test           # fixes/modifies tests
```

# Context

This is another step forward on addressing https://github.com/IntersectMBO/cardano-cli/issues/882.

Previous related PR: https://github.com/IntersectMBO/cardano-cli/pull/915

This only adds checks to commands inside `governance drep`. There are several other places where to add the check, but better to go step by step.

# How to trust this PR

I have adapted test for all commands but one, to validate the new functionality, and that gives me a lot of confidence. I would try to check that the code style is good, I didn't introduce any bugs, the parameter structure and names make sense, and the help is comprehensive and comprehensible.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Types/Common.hs [hunk 1]
```diff
@@ -645,3 +651,12 @@ data MustCheckHash a
   = CheckHash
   | TrustHash
   deriving (Eq, Show)
+
+data PotentiallyCheckedAnchor anchorType anchor
+  = PotentiallyCheckedAnchor
+  { pcaAnchor :: anchor
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Types/Common.hs]: The `anchor` type parameter seems redundant here. AFAIU you added that because sometimes you have `Anchor StandardCrypto` and sometimes `Anchor (EraCrypto (ShelleyLedger Era StandardCrypto))`, correct?

From the CLI point, it will be always `EraCrypto _ ~ StandardCrypto`.


Can you try to remove `anchor` parameter and try to use `StandardCrypto` everywhere? Are you expecting to support other `anchor` types here?

File: cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs [hunk 2]
```diff
@@ -53,6 +56,10 @@ data GovernanceCmdError
   | -- Legacy - remove me after cardano-cli transitions to new era based structure
     GovernanceCmdMIRCertNotSupportedInConway
   | GovernanceCmdGenesisDelegationNotSupportedInConway
+  | GovernanceCmdMismatchedDRepMetadataHashError
+      !(L.SafeHash L.StandardCrypto L.AnchorData)
+      !(L.SafeHash L.StandardCrypto L.AnchorData)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs]: Can you add a comments which one is expected and which one is actual?

File: cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs [hunk 3]
```diff
@@ -6,15 +6,23 @@ module Cardano.CLI.Types.Errors.RegistrationError
...
+import           Control.Exception (displayException)
+
 data RegistrationError
   = RegistrationReadError !(FileError InputDecodeError)
   | RegistrationWriteFileError !(FileError ())
   | RegistrationStakeCredentialError !StakeCredentialError
   | RegistrationStakeError !StakeAddressRegistrationError
+  | RegistrationMismatchedDRepMetadataHashError
+      !(L.SafeHash L.StandardCrypto L.AnchorData)
+      !(L.SafeHash L.StandardCrypto L.AnchorData)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs]: Can you add haddocks, which one is expected and which one is actual?

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs [hunk 4]
```diff
@@ -67,7 +73,13 @@ data GovernanceDRepUpdateCertificateCmdArgs era
   = GovernanceDRepUpdateCertificateCmdArgs
   { eon :: !(ConwayEraOnwards era)
   , drepHashSource :: !DRepHashSource
-  , mAnchor :: !(Maybe (L.Anchor (L.EraCrypto (ShelleyLedgerEra era))))
+  , mPotentiallyCheckedAnchor
+      :: !( Maybe
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: Maybe is lazy so `!` won't buy you much here anyway. But 👍🏻

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs [hunk 5]
```diff
@@ -171,3 +188,28 @@ runGovernanceDRepMetadataHashCmd
...
       . serialiseToRawBytesHex
       $ metadataHash
+
+-- | Check the hash of the anchor data against the hash in the anchor if
+-- checkHash is set to CheckHash.
+carryHashChecks
+  :: (FetchURLError -> error)
+  -- ^ Function that takes a FetchURLError and returns the error type of the caller
+  -> (L.SafeHash L.StandardCrypto L.AnchorData -> L.SafeHash L.StandardCrypto L.AnchorData -> error)
+  -- ^ Function that takes the expected and actual hashes and returns the mismatch hash error type of the caller
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: The error handling in this function type is too noisy imo. I think this should be a separate error datatype instead of using function arguments.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs [hunk 6]
```diff
@@ -111,15 +111,18 @@ pRegistrationCertificateCmd era = do
...
         <*> pKeyRegistDeposit
-        <*> pDRepMetadata
+        <*> optional
+          ( pPotentiallyCheckedAnchorData
+              (pMustCheckHash drepMetadataHashCheckInfo)
+              pDRepMetadata
+          )
         <*> pOutputFile
 
-pDRepMetadata :: Parser (Maybe (L.Anchor L.StandardCrypto))
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs:117]: Nice change :+1: It's better to have the caller do the call to `optional` indeed

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs [hunk 7]
```diff
@@ -171,3 +188,28 @@ runGovernanceDRepMetadataHashCmd
...
+  -> PotentiallyCheckedAnchor DRepMetadataUrl (L.Anchor L.StandardCrypto)
+  -- ^ The information about anchor data and whether to check the hash (see 'PotentiallyCheckedAnchor')
+  -> ExceptT error IO ()
+carryHashChecks errorAdaptor hashMismatchError potentiallyCheckedAnchor =
+  let anchor = pcaAnchor potentiallyCheckedAnchor
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: I'd put that in a `where` clause to reduce the level of indentation in the function.

Note that I usually refrain from subjective suggestions like this, but in this case, I find the body quite hard to read with the number of indentations (in a small number of lines).

File: cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs [hunk 8]
```diff
@@ -114,5 +121,13 @@ instance Error GovernanceCmdError where
       "MIR certificates are not supported in Conway era onwards."
     GovernanceCmdGenesisDelegationNotSupportedInConway ->
       "Genesis delegation is not supported in Conway era onwards."
+    GovernanceCmdMismatchedDRepMetadataHashError expectedHash actualHash ->
+      "DRep metadata Hashes do not match!"
+        <> "\nExpected:"
+          <+> pretty (show (L.extractHash expectedHash))
+        <> "\n  Actual:"
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs]:
```suggestion
        <> "\nActual:"
```

Bit weird to indent the `Actual` case, whereas you don't indent the `Expected` case?

File: cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs [hunk 9]
```diff
@@ -27,3 +35,11 @@ instance Error RegistrationError where
       "Cannot read stake credential: " <> prettyError e
     RegistrationStakeError e ->
       "Stake address registation error: " <> prettyError e
+    RegistrationMismatchedDRepMetadataHashError expectedHash actualHash ->
+      "DRep metadata Hashes do not match!"
+        <> "\nExpected:"
+          <+> pretty (show (L.extractHash expectedHash))
+        <> "\n  Actual:"
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs:45]:
```suggestion
        <> "\nActual:"
```

Bit weird to indent the `Actual` case, whereas you don't indent the `Expected` case?

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs [hunk 10]
```diff
@@ -67,7 +73,13 @@ data GovernanceDRepUpdateCertificateCmdArgs era
   = GovernanceDRepUpdateCertificateCmdArgs
   { eon :: !(ConwayEraOnwards era)
   , drepHashSource :: !DRepHashSource
-  , mAnchor :: !(Maybe (L.Anchor (L.EraCrypto (ShelleyLedgerEra era))))
+  , mPotentiallyCheckedAnchor
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: I'd keep the field name `mAnchor` since the type already gives information about the need to check the hash. Long field names are hard to parse IMHO.

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs (hunk 7 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/2e8b2b84fb9808012aa18ad1b702ea6223aa6c37

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs (hunk 5 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/0b03d1d8fcb63f66883246f2a4081e424c1cd5c0

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs (hunk 10 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/773ad8cedb7b6f5676e6d7f373be15633fae5abd

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs (hunk 4 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: Ah, makes sense. I was just following the pattern, I will remove the bang

Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/2825cc930b6d85d080524baf468e21da232800e1

File: cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs (hunk 8 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs]: Yes, a bit, but it is on purpouse to align the hashes so that they are easier to compare visually. I could keep the "Actual:" to the left and only align the hash, but I think it probably looks nicer like this

File: cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs (hunk 9 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs:45]: See https://github.com/IntersectMBO/cardano-cli/pull/916#discussion_r1781299176

File: cardano-cli/src/Cardano/CLI/Types/Common.hs (hunk 1 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/Common.hs]: It was more for the URL than the hash. For example, in the other PR the type was:
```haskell
  , proposalUrl :: !ProposalUrl
  , proposalHash :: !(L.SafeHash L.StandardCrypto L.AnchorData)
```
So I was opening the door for other packagings. But it may be that this doesn't happen inside a `Maybe`.
I will try to instantiate for now, and generalize if needed later, that is good practice

Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/Common.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/4d736e59bd7f062fb99da1691dfa05dd1b210c80

File: cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs (hunk 2 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/4ab4a52e4680bd14ea9744981f1722705081f1fd

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs (hunk 6 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs:117]: Thanks 🙂

File: cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs (hunk 3 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/Errors/RegistrationError.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/6a0290eb88555536b54dab922d98bcabdb827d02

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs [hunk 11]
```diff
@@ -35,6 +40,11 @@ drepRegistrationCertSem :: FileSem
 drepRegistrationCertSem = newFileSem "test/cardano-cli-golden/files/golden/governance/drep/drep_registration_certificate.json"
 {-# NOINLINE drepRegistrationCertSem #-}
 
+-- | Semaphore protecting against locked file error, when running properties concurrently.
+drepRegistrationCertSem2 :: FileSem
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Why bother to run these tests  concurrently and introduce semaphores? They're golden tests so they should run quickly on its own. Do they take that much longer when run sequentially?

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs [hunk 12]
```diff
@@ -204,29 +214,50 @@ hprop_golden_governance_drep_metadata_hash_cip119 = propertyOnce . H.moduleWorks
 
   H.diffFileVsGoldenFile outputDRepMetadataHashCip119 goldenDRepMetadataHashCip119
 
+hprop_golden_governance_drep_registration_certificate_vkey_file_wrong_hash_fails :: Property
+hprop_golden_governance_drep_registration_certificate_vkey_file_wrong_hash_fails =
+  propertyOnce . expectFailure . H.moduleWorkspace "tmp" $ \tempDir ->
+    base_golden_governance_drep_registration_certificate_vkey_file
+      ('a' : drop 1 exampleAnchorDataHash)
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Why? Where is the comment explaining why you are randomly prepending `'a'`  here? Think about the engineer that has to read this code months down the line and has to understand what is happening here.

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs [hunk 13]
```diff
@@ -204,29 +214,50 @@ hprop_golden_governance_drep_metadata_hash_cip119 = propertyOnce . H.moduleWorks
...
 
-  void $ execCardanoCLI
-    [ "conway", "governance", "drep", "registration-certificate"
-    , "--drep-verification-key-file", drepVKeyFile
-    , "--key-reg-deposit-amt", "0"
-    , "--drep-metadata-url", "dummy-url"
-    , "--drep-metadata-hash", "52e69500a92d80f2126c836a4903dc582006709f004cf7a28ed648f732dff8d2"
-    , "--out-file", outFile
-    ]
+  let relativeUrl = ["ipfs", exampleAnchorDataIpfsHash]
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: This is no longer really a golden file test. This should be tested separately, not squeezed into the existing golden tests. It belongs in `cardano-cli-test`.

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs [hunk 14]
```diff
@@ -236,16 +267,17 @@ hprop_golden_governance_drep_registration_certificate_id_hex = propertyOnce . H.
     , "--key-reg-deposit-amt", "0"
     , "--drep-metadata-url", "dummy-url"
     , "--drep-metadata-hash", "52e69500a92d80f2126c836a4903dc582006709f004cf7a28ed648f732dff8d2"
+    , "--trust-drep-metadata-hash"
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Not checking the hash should be a silent default. You need to opt into this behaviour.

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs [hunk 15]
```diff
@@ -273,30 +306,54 @@ hprop_golden_governance_drep_registration_certificate_script_hash = propertyOnce
...
-    , "--drep-metadata-url", "dummy-url"
-    , "--drep-metadata-hash", "52e69500a92d80f2126c836a4903dc582006709f004cf7a28ed648f732dff8d2"
-    , "--out-file", outFile
-    ]
+  let relativeUrl = ["ipfs", exampleAnchorDataIpfsHash]
+  serveFilesWhile
+    [(relativeUrl, exampleAnchorDataPathGolden)]
+    ( \port -> do
+        void $
+          execCardanoCLIWithEnvVars
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Same here, this is no longer really a golden file test.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs [hunk 16]
```diff
@@ -3628,6 +3637,15 @@ constitutionHashCheckInfo =
     , urlParamName = "--constitution-url"
     }
 
+drepMetadataHashCheckInfo :: HashCheckParamInfo DRepMetadataUrl
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs]: This needs to be rectified in `cardano-cli` before this PR is merged. This is poorly typed and an unnecessary intermediate representation.

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs [hunk 17]
```diff
@@ -171,3 +206,31 @@ runGovernanceDRepMetadataHashCmd
       . writeByteStringOutput mOutFile
       . serialiseToRawBytesHex
       $ metadataHash
+
+data HashCheckError
+  = HashMismatchError
+      (L.SafeHash L.StandardCrypto L.AnchorData)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Where are the comments? Make it easy for the reader to know what the expected hash is.

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs [hunk 18]
```diff
@@ -150,10 +170,25 @@ runGovernanceDRepUpdateCertificateCmd
...
       drepCredential <- modifyError GovernanceCmdKeyReadError $ readDRepCredential drepHashSource
-      let updateCertificate = makeDrepUpdateCertificate (DRepUpdateRequirements w drepCredential) mAnchor
+      let updateCertificate =
+            makeDrepUpdateCertificate
+              (DRepUpdateRequirements w drepCredential)
+              (pcaAnchor <$> mAnchor)
       firstExceptT GovernanceCmdTextEnvWriteError . newExceptT $
         writeFileTextEnvelope outFile (Just "DRep Update Certificate") updateCertificate
+   where
+    hashCheckToGovernanceCmdError :: ExceptT HashCheckError IO () -> ExceptT GovernanceCmdError IO ()
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Why not add a new constructor to `GovernanceCmdError` to encapsulate `HashCheckError`?  That is the usual pattern why are we deviating from this now?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 14 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Based on [this discussion](https://x.com/carloslodelar/status/1833506267102400763), it was my understanding that this was the consensus. Do you mean we should do it differently in this particular case?

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 11 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: To be honest, I just copied the previous line, didn't think much about it. Probably they are pretty fast, like you say, we can make them sequential as far as I am concerned

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 12 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/916/commits/541fae33adf83831e97e5e104df82ec9f2baba33

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 13 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/916/commits/e805935a1d8b0dd27b47b479fd03ce52bd4967e8

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 15 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/916/commits/541fae33adf83831e97e5e104df82ec9f2baba33

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs (hunk 17 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/916/commits/e4481cb889d795faf882b5ed62492b94b30b0dbd

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs (hunk 18 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/916/commits/9c82cfdada338d3fefc0874fff33bf1261ee3883

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs (hunk 16 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/916/commits/710ef81b87a51c6c9fa830ad964846413ee8e6a5

**Comment by palas** [comment]
Rebased to main. Before rebase: https://github.com/IntersectMBO/cardano-cli/tree/backup/add-hash-validation3

File: cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs (hunk 8 above)
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/Types/Errors/GovernanceCmdError.hs]: Ah I see, makes sense :+1: Thanks @palas for the explanation :pray:

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 14 above)
Comment (Jimbo4350) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Having it in the help text is enough to convey to users they can use this feature if they want. It's unnecessary to make users explicitly opt out.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs [hunk 19]
```diff
@@ -3610,23 +3610,44 @@ pMustCheckHash
...
+  -> Parser (PotentiallyCheckedAnchor anchorDataType)
+pPotentiallyCheckedAnchorData mustCheckHash anchorData =
+  PotentiallyCheckedAnchor
+    <$> anchorData
+    <*> mustCheckHash
+
+pMustCheckProposalHash :: Parser (MustCheckHash ProposalUrl)
+pMustCheckProposalHash =
+  pMustCheckHash
+    HashCheckParamInfo
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs]: This has to be fixed on `main` first before I approve this PR. I do not want this being propagated.

Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs:3625]: Won't merging this PR fix it in main?

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 14 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Got it. I initially followed our earlier discussion and still think that approach made sense, but I’m fine with changing it. Let me know if you want to chat more, or I’ll go ahead with the updates

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs (hunk 19 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/921

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 14 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/baafefd637584bcf2faf7d8549237f7718b364a9

File: cardano-cli/test/cardano-cli-golden/files/golden/help/latest_governance_drep_registration-certificate.cli [hunk 20]
```diff
@@ -25,5 +26,9 @@ Available options:
   --drep-metadata-url TEXT DRep anchor URL
   --drep-metadata-hash HASH
                            DRep anchor data hash.
+  --check-drep-metadata-hash
+                           Check the DRep metadata hash (from
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/files/golden/help/latest_governance_drep_registration-certificate.cli]: "A check that calculates a hash from the data and compares it to the command line specified hash" ? This wording could be improved but what does "Check" mean? We need to be clearer.

File: cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Hash.hs [hunk 21]
```diff
@@ -51,6 +53,18 @@ exampleAnchorDataIpfsHash, exampleAnchorDataIpfsHash2 :: String
 exampleAnchorDataIpfsHash = "QmbL5EBFJLf8DdPkWAskG3Euin9tHY8naqQ2JDoHnWHHXJ"
 exampleAnchorDataIpfsHash2 = "QmdTJ4PabgSabg8K1Z4MNXnSVM8bjJnAikC3rVWfPVExQj"
 
+-- | Tamper with the base16 hash by adding one to the first character
+tamperBase16Hash :: String -> Maybe String
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Hash.hs:57]: This could be a hedgehog generator.

File: cardano-cli/test/cardano-cli-test/Test/Cli/Governance/DRep.hs [hunk 22]
```diff
@@ -82,3 +91,121 @@ hprop_governance_drep_update_certificate_vkey_file =
...
+      , "--verification-key-file"
+      , drepVKeyFile
+      , "--signing-key-file"
+      , drepSKeyFile
+      ]
+
+  outFile <- H.noteTempFile tempDir "drep-reg-cert.txt"
+
+  let relativeUrl = ["ipfs", exampleAnchorDataIpfsHash]
+  serveFilesWhile
```
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-test/Test/Cli/Governance/DRep.hs:134]: A comment here about what this is doing would be useful.

**Review by Jimbo4350 — APPROVED** [review]
LGTM! A few comments but it can be merged :+1:

File: cardano-cli/test/cardano-cli-golden/files/golden/help/latest_governance_drep_registration-certificate.cli (hunk 20 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/files/golden/help/latest_governance_drep_registration-certificate.cli]: Maybe the confusing bit is that it seems like a noun "check" while I meant the imperative of the verb "to check". I will write "verify that", that should be clearer

File: cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Hash.hs (hunk 21 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Hash.hs:57]: I will leave it for now, because that would be more for testing the actual hash comparison, which was implemented by someone else. Maybe in another PR

Comment (palas) [reply cardano-cli/test/cardano-cli-test-lib/Test/Cardano/CLI/Hash.hs:57]: Maybe it is a good tool to add to hedgehog extras though
